File: internal/db/models/vm/vm.go
// SPDX-License-Identifier: BSD-2-Clause
//
// Copyright (c) 2025 The FreeBSD Foundation.
//
// This software was developed by Hayzam Sherif <hayzam@alchemilla.io>
// of Alchemilla Ventures Pvt. Ltd. <hello@alchemilla.io>,
// under sponsorship from the FreeBSD Foundation.

package vmModels

import (
	"fmt"
	"time"

	networkModels "github.com/alchemillahq/sylve/internal/db/models/network"
	"github.com/digitalocean/go-libvirt"
	"gorm.io/gorm"
)

func (Storage) TableName() string {
	return "vm_storages"
}

type VMStorageType string

const (
	VMStorageTypeRaw       VMStorageType = "raw"
	VMStorageTypeZVol      VMStorageType = "zvol"
	VMStorageTypeDiskImage VMStorageType = "image"
)

type VMStorageEmulationType string

const (
	VirtIOStorageEmulation VMStorageEmulationType = "virtio-blk"
	AHCIHDStorageEmulation VMStorageEmulationType = "ahci-hd"
	AHCICDStorageEmulation VMStorageEmulationType = "ahci-cd"
	NVMEStorageEmulation   VMStorageEmulationType = "nvme"
)

type TimeOffset string

const (
	TimeOffsetUTC   TimeOffset = "utc"
	TimeOffsetLocal TimeOffset = "localtime"
)

type VMStorageDataset struct {
	ID   uint   `gorm:"primaryKey" json:"id"`
	Pool string `json:"pool"`
	Name string `json:"name"`
	GUID string `json:"guid"`
}

func (VMStorageDataset) TableName() string {
	return "vm_storage_datasets"
}

type Storage struct {
	ID   uint          `gorm:"primaryKey" json:"id"`
	Type VMStorageType `json:"type"`

	Name         string `json:"name"`
	DownloadUUID string `json:"uuid"`

	Pool string `json:"pool"`

	DatasetID *uint            `json:"datasetId" gorm:"column:dataset_id"`
	Dataset   VMStorageDataset `json:"dataset" gorm:"foreignKey:DatasetID;references:ID"`

	Size      int64                  `json:"size"`
	Emulation VMStorageEmulationType `json:"emulation"`

	RecordSize   int `json:"recordSize"`
	VolBlockSize int `json:"volBlockSize"`

	BootOrder int  `json:"bootOrder"`
	VMID      uint `json:"vmId" gorm:"index"`
}

func (Network) TableName() string {
	return "vm_networks"
}

type Network struct {
	ID  uint   `gorm:"primaryKey" json:"id"`
	MAC string `json:"mac"`

	MacID      *uint                 `json:"macId" gorm:"column:mac_id"`
	AddressObj *networkModels.Object `json:"macObj" gorm:"foreignKey:MacID"`

	SwitchID   uint   `json:"switchId" gorm:"index;not null"`
	SwitchType string `json:"switchType" gorm:"index;not null;default:standard"`

	StandardSwitch *networkModels.StandardSwitch `gorm:"-" json:"-"`
	ManualSwitch   *networkModels.ManualSwitch   `gorm:"-" json:"-"`

	Emulation string `json:"emulation"`
	VMID      uint   `json:"vmId" gorm:"index"`
}

func (n *Network) AfterFind(tx *gorm.DB) error {
	switch n.SwitchType {
	case "standard":
		var s networkModels.StandardSwitch
		if err := tx.First(&s, n.SwitchID).Error; err != nil {
			return fmt.Errorf("load standard switch %d: %w", n.SwitchID, err)
		}
		n.StandardSwitch = &s
	case "manual":
		var m networkModels.ManualSwitch
		if err := tx.First(&m, n.SwitchID).Error; err != nil {
			return fmt.Errorf("load manual switch %d: %w", n.SwitchID, err)
		}
		n.ManualSwitch = &m
	default:
		return fmt.Errorf("unknown switch type: %s", n.SwitchType)
	}
	return nil
}

type VMStats struct {
	ID          uint    `gorm:"primaryKey" json:"id"`
	VMID        uint    `json:"vmId" gorm:"index"`
	CPUUsage    float64 `json:"cpuUsage"`
	MemoryUsage float64 `json:"memoryUsage"`
	MemoryUsed  float64 `json:"memoryUsed"`

	CreatedAt time.Time `json:"createdAt" gorm:"autoCreateTime"`
}

func (VMStats) TableName() string {
	return "vm_stats"
}

func (s VMStats) GetID() uint {
	return s.ID
}

func (s VMStats) GetCreatedAt() time.Time {
	return s.CreatedAt
}

type VMCPUPinning struct {
	ID   uint `gorm:"primaryKey" json:"id"`
	VMID uint `json:"vmId" gorm:"index"`

	HostSocket int   `json:"hostSocket"`
	HostCPU    []int `json:"hostCpu" gorm:"serializer:json;type:json"`
}

type VM struct {
	ID          uint   `gorm:"primaryKey" json:"id"`
	Name        string `json:"name"`
	Description string `json:"description"`
	RID         uint   `json:"rid" gorm:"column:rid;not null;uniqueIndex;"`

	CPUSockets int `json:"cpuSockets"`
	CPUCores   int `json:"cpuCores"`
	CPUThreads int `json:"cpuThreads"`

	CPUPinning []VMCPUPinning `json:"cpuPinning" gorm:"foreignKey:VMID;references:ID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`

	RAM int `json:"ram"`

	TPMEmulation     bool `json:"tpmEmulation"`
	ShutdownWaitTime int  `json:"shutdownWaitTime" gorm:"default:10"`

	Serial bool `json:"serial" gorm:"default:false"`

	VNCEnabled    bool   `json:"vncEnabled" gorm:"default:true"`
	VNCPort       int    `json:"vncPort"`
	VNCPassword   string `json:"vncPassword"`
	VNCResolution string `json:"vncResolution"`
	VNCWait       bool   `json:"vncWait"`

	StartAtBoot bool       `json:"startAtBoot"`
	StartOrder  int        `json:"startOrder"`
	WoL         bool       `json:"wol" gorm:"default:false"`
	TimeOffset  TimeOffset `json:"timeOffset" gorm:"default:'utc'"`

	Storages   []Storage `json:"storages" gorm:"foreignKey:VMID;references:ID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`
	Networks   []Network `json:"networks" gorm:"foreignKey:VMID;references:ID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`
	PCIDevices []int     `json:"pciDevices" gorm:"serializer:json;type:json"`

	ACPI bool `json:"acpi"`
	APIC bool `json:"apic"`

	Stats []VMStats           `json:"-" gorm:"foreignKey:VMID;references:ID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`
	State libvirt.DomainState `json:"state" gorm:"-"`

	CloudInitData            string `json:"cloudInitData" gorm:"type:text"`
	CloudInitMetaData 		 string `json:"cloudInitMetaData" gorm:"type:text"`
	CloudInitNetworkConfig   string `json:"cloudInitNetworkConfig" gorm:"type:text"`
	IgnoreUMSR        		 bool   `json:"ignoreUMSR" gorm:"default:false"`

	CreatedAt time.Time `json:"createdAt" gorm:"autoCreateTime"`
	UpdatedAt time.Time `json:"updatedAt" gorm:"autoUpdateTime"`

	StartedAt *time.Time `json:"startedAt" gorm:"default:null"`
	StoppedAt *time.Time `json:"stoppedAt" gorm:"default:null"`
}

File: web/src/lib/types/vm/vm.ts
import { z } from 'zod/v4';
import { NetworkObjectSchema } from '../network/object';

export interface CPUPin {
    socket: number;
    cores: number[];
}

export interface CreateData {
    name: string;
    node: string;
    id: number;
    description: string;
    storage: {
        type: string;
        pool: string;
        size: number;
        emulation: string;
        iso: string;
    };
    network: {
        switch: string;
        mac: string;
        emulation: string;
    };
    hardware: {
        sockets: number;
        cores: number;
        threads: number;
        memory: number;
        passthroughIds: number[];
        pinnedCPUs: CPUPin[];
        isPinningOpen: boolean;
    };
    advanced: {
        serial: boolean;
        vncPort: number;
        vncPassword: string;
        vncWait: boolean;
        vncResolution: string;
        startAtBoot: boolean;
        bootOrder: number;
        tpmEmulation: boolean;
        timeOffset: 'utc' | 'localtime';
        cloudInit: {
            enabled: boolean;
            data: string;
            metadata: string;
            networkConfig: string;
        };
        ignoreUmsrs: boolean;
    };
}

export type VMStorageType = 'raw' | 'zvol' | 'image';
export type VMStorageEmulationType = 'virtio-blk' | 'ahci-hd' | 'ahci-cd' | 'nvme';

export const VMStorageDatasetSchema = z.object({
    id: z.number().int(),
    pool: z.string(),
    name: z.string(),
    guid: z.string()
});

export const VMStorageSchema = z.object({
    id: z.number().int(),
    vmId: z.number().int().optional(),
    name: z.string().optional(),
    type: z.enum(['raw', 'zvol', 'image']),
    uuid: z.string().optional(),
    datasetId: z.number().int().nullable(),
    dataset: VMStorageDatasetSchema.nullable(),
    size: z.number().int(),
    emulation: z.enum(['virtio-blk', 'ahci-hd', 'ahci-cd', 'nvme']),
    recordSize: z.number().int().optional(),
    volBlockSize: z.number().int().optional(),
    bootOrder: z.number().int().optional()
});

export const VMNetworkSchema = z.object({
    id: z.number().int(),
    mac: z.string(),
    macId: z.number().int().optional(),
    macObj: NetworkObjectSchema.optional(),
    switchId: z.number().int(),
    switchType: z.enum(['standard', 'manual']),
    emulation: z.string(),
    vmId: z.number().int().optional()
});

export const VMCPUPinningSchema = z.object({
    id: z.number().int(),
    vmId: z.number().int(),
    hostSocket: z.number().int(),
    hostCpu: z.array(z.number().int())
});

export enum DomainState {
    DomainNostate = 0,
    DomainRunning = 1,
    DomainBlocked = 2,
    DomainPaused = 3,
    DomainShutdown = 4,
    DomainShutoff = 5,
    DomainCrashed = 6,
    DomainPmsuspended = 7
}

export const DomainStateSchema = z.enum(DomainState);

export const VMSchema = z.object({
    id: z.number().int(),
    name: z.string(),
    description: z.string(),
    rid: z.number().int(),
    cpuSockets: z.number().int(),
    cpuCores: z.number().int(),
    cpuThreads: z.number().int(),
    ram: z.number().int(),
    serial: z.boolean(),
    vncEnabled: z.boolean(),
    vncPort: z.number().int(),
    vncPassword: z.string(),
    vncResolution: z.string(),
    vncWait: z.boolean(),
    startAtBoot: z.boolean(),
    startOrder: z.number().int(),
    wol: z.boolean(),
    timeOffset: z.enum(['utc', 'localtime']),
    state: DomainStateSchema,
    storages: z.array(VMStorageSchema),
    networks: z.array(VMNetworkSchema),
    pciDevices: z.union([z.array(z.number().int()), z.null()]),
    cpuPinning: z.union([z.array(VMCPUPinningSchema), z.null()]),
    shutdownWaitTime: z.number().int(),
    cloudInitData: z.string().nullable(),
    cloudInitMetaData: z.string().nullable(),
    cloudInitNetworkConfig: z.string().nullable(),
    ignoreUMSR: z.boolean(),
    tpmEmulation: z.boolean(),

    createdAt: z.string(),
    updatedAt: z.string(),

    startedAt: z.string().nullable(),
    stoppedAt: z.string().nullable()
});

export const VMStatSchema = z.object({
    vmId: z.number().int().default(0),
    cpuUsage: z.number().default(0),
    memoryUsage: z.number().default(0),
    memoryUsed: z.number().default(0),
    createdAt: z.string().default(new Date(0).toISOString())
});

export const VMDomainSchema = z.object({
    id: z.number().int(),
    uuid: z.string(),
    name: z.string(),
    status: z.string()
});

export const SimpleVmSchema = z.object({
    id: z.number().int(),
    name: z.string(),
    rid: z.number().int(),
    vncPort: z.number(),
    state: DomainStateSchema
});

export type VM = z.infer<typeof VMSchema>;
export type VMStorage = z.infer<typeof VMStorageSchema>;
export type VMNetwork = z.infer<typeof VMNetworkSchema>;
export type VMDomain = z.infer<typeof VMDomainSchema>;
export type VMStat = z.infer<typeof VMStatSchema>;
export type SimpleVm = z.infer<typeof SimpleVmSchema>;
